recirq/quantum_chess/circuit_transformer.py

```diff
@@ -305,6 +307,391 @@ def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
+          lg: A logical qubits graph.
+        """
+        mapping = defaultdict()
+
+        # The circuit doesn't contain any 2-qubit gates, return an empty
+        # mapping.
+        if not lg:
+            return mapping
+
+        pg_center = self.find_graph_center(pg)
+        lg_center = self.find_graph_center(lg)
+        mapping[lg_center] = pg_center
+
+        traversal_order = self.traverse(lg, lg_center)
+
+        while traversal_order:
+            lq = traversal_order.popleft()
+            if lq == lg_center:
+                continue
+            pq = None
+            reference_qubits = self.find_reference_qubits(mapping, lg, lq)
+            ref_q = reference_qubits[0]
+            candidate_qubits = self.find_candidate_qubits(mapping.values(), pg, ref_q)
+            if len(reference_qubits) > 1:
+                # For each reference location, find the shortest path distance
+                # to each of the candidate qubits. Only keep the nearest
+                # candidate qubits with smallest distance.
+                for ref_q in reference_qubits[1:]:
+                    distances = defaultdict(list)
+                    for cand_q in candidate_qubits:
+                        d = self.find_shortest_path_distance(pg, ref_q, cand_q)
+                        distances[d].append(cand_q)
+                    nearest_candidate_qubits = None
+                    min_dist = math.inf
+                    for dist in distances:
+                        if dist < min_dist:
+                            min_dist = dist
+                            nearest_candidate_qubits = distances[dist]
+                    candidate_qubits = nearest_candidate_qubits
+                    if candidate_qubits == 1:
+                        pq = candidate_qubits[0]
+                        break
+            # If there are still more than one candidate qubit at this point,
+            # choose the one with the closest degree to the logical qubit.
+            if len(candidate_qubits) > 1:
+                lq_degree = len(lg[lq])
+                min_diff = math.inf
+                for cand_q in candidate_qubits:
+                    cand_q_degree = len(pg[cand_q])
+                    diff = abs(cand_q_degree - lq_degree)
+                    if diff < min_diff:
+                        min_diff = diff
+                        pq = cand_q
+            mapping[lq] = pq
+
+        return mapping
+
+    def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
+        """Returns a transformed circuit.
+
+        The transformed circuit satisfies all physical adjacency constraints
+        of the device.
+
+        Args:
+          circuit: The circuit to transform.
+        """
+        pg = self.build_physical_qubits_graph()
+        lg = self.build_logical_qubits_graph(circuit)
+        initial_mapping = self.calculate_initial_mapping(pg, lg)
+        # TODO: return the transformed circuit here instead.
+        return initial_mapping
```

Comment:
I think this doesn't type -- the initial mapping is a dict, but the return type is Circuit. We could transform_qubits() on the input circuit to get a circuit representing the initial mapping.

And if we did that, we could run the circuit_transformer_test cases on the DynamicLookAheadHeuristicCircuitTransformer as well as the existing one to see what happens -- I have a feeling that the initial mapping will be able to handle the small circuits in those test cases

Reply:
the initial mapping algorithm ignores 1-qubit gates, and for some reason i had thought the swap update algorithm would take care of including them. but yah, i just updated the logic to handle 1-qubit gates (we're considering each as a single connected component).

i added tests for DynamicLookAheadHeuristicCircuitTransformer to the first three cases in circuit_transformer_test.py; since the heuristic used in this mapping algorithm is not exhaustive like in ConnectivityHeuristicCircuitTransformer, the last three cases will be flaky until we implement the swap update algorithm.

Reply:
I see. That is a bit unfortunate that even those small examples would require additional swaps, but good that we can add some extra test coverage!